### PR TITLE
feat: add workspace name in model breadcrumb

### PR DESCRIPTION
**Pull request by keita-determined**
## Description

[WEB-806]

https://user-images.githubusercontent.com/109113616/216693028-14d509ea-3e6c-4375-9236-b426c2a7918b.mov

## Test Plan
- Breadcrumb: `{Workspace name}/Model Registry/{model name}/{model version}?` (`?` means optional)
- Check if user can go back to parent item by clicking breadcrumb including workspace
  - If user clicks `Model Registry`, it leads to the global model registry page

## Commentary (optional)
I split the URL for global model registry and workspace model registry to make things clear. 

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[WEB-806]: https://determinedai.atlassian.net/browse/WEB-806?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by keita-determined** [comment]
@ClaireNeveu 
question about [this](https://determinedai.atlassian.net/browse/WEB-806?focusedCommentId=19039): 

for the global model registry, there's `Model Registry` in breadcrumb (first image). I think it does make sense to have.
Meanwhile, there's `Model Registry` in workspace model registry (second image). i think it doesn't make sense to have. Currently, I leave as it is, but it leads to global model registry page. Should we remove it for workspace model registry? 

 In summary, 
- `Model Registry/{model name}/{model version name}` for global model registry
- `{Workspace name}/{model name}/{model version name}` for workspace model registry

Does that make sense? 

image

image

P.S. Are there good names for global model registry and workspace model registry?

File: webui/react/src/routes/routes.ts [hunk 1]
```diff
@@ -9,6 +9,24 @@ const routes: RouteConfig[] = [
...
+    needAuth: true,
+    path: '/workspaces/:workspaceId/models/:modelId/versions/:versionNum',
+    title: 'Version Details',
+  },
+  {
+    id: 'modelDetails',
+    needAuth: true,
+    path: '/workspaces/:workspaceId/models/:modelId',
+    title: 'Model Details',
+  },
```
Comment (keita-determined) [inline webui/react/src/routes/routes.ts:29]: I split the URL for global model registry and workspace model registry to make things clear.

**Comment by ClaireNeveu** [comment]
I like having model registry within the workspace since the objects contained in a workspace are heterogenous. You can have that breadcrumb link directly to the appropriate tab.

 Also worth noting here that the breadcrumb trail should always be the same no matter how you got to it. If you use the global model registry we should still be going to the workspace-specific page when you click into a model.

File: webui/react/src/routes/routes.ts (hunk 1 above)
Comment (ClaireNeveu) [reply webui/react/src/routes/routes.ts:29]: We don't actually want separate urls, there should only be one canonical url for a model and model version.

File: webui/react/src/components/Table/Table.tsx [hunk 2]
```diff
@@ -200,20 +200,36 @@ export const experimentProgressRenderer: ExperimentRenderer = (_, record) => {
...
-    <div style={{ paddingInline: 4 }}>
-      <Icon name="model" size="medium" />
-    </div>
-    <Link path={paths.modelDetails(String(record.id))}>{value}</Link>
-  </Space>
-);
+export const modelNameRenderer = (
+  value: string,
+  record: ModelItem,
+  workspaceId?: number,
```
Comment (mapmeld) [inline webui/react/src/components/Table/Table.tsx:206]: from yesterday we will record `model.workspace_id` from the API, would you want to use that and remove workspaceId as a separate param here?

File: webui/react/src/pages/ModelVersionDetails.tsx [hunk 3]
```diff
@@ -96,6 +116,10 @@ const ModelVersionDetails: React.FC = () => {
     }
   }, [basePath, navigate, tab, tabKey]);
 
+  useEffect(() => {
+    fetchWorkspace();
```
Comment (mapmeld) [inline webui/react/src/pages/ModelVersionDetails.tsx:120]: fetchWorkspace / setWorkspace can also be avoided if you use workspace ID from the new `modelVersion.model.workspaceId`

File: webui/react/src/components/Table/Table.tsx (hunk 2 above)
Comment (keita-determined) [reply webui/react/src/components/Table/Table.tsx:206]: Thanks for the info!

File: webui/react/src/pages/ModelVersionDetails.tsx (hunk 3 above)
Comment (keita-determined) [reply webui/react/src/pages/ModelVersionDetails.tsx:120]: i fetch workspace to get workspace name since `ModelItem` doesnt have workspace name.

File: master/static/srv/get_model_version.sql [hunk 4]
```diff
@@ -5,7 +5,7 @@ WITH mv AS (
     WHERE model_id = $1 AND model_versions.version = $2
 ),
 m AS (
-  SELECT m.id, m.name, m.description, m.notes, m.metadata, m.creation_time, m.last_updated_time, array_to_json(m.labels) AS labels, u.username, m.user_id, m.archived, COUNT(mv.version) as num_versions
+  SELECT m.id, m.name, m.description, m.notes, m.metadata, m.creation_time, m.last_updated_time, array_to_json(m.labels) AS labels, u.username, m.user_id, m.archived, COUNT(mv.version) as num_versions, m.workspace_id
```
Comment (keita-determined) [inline master/static/srv/get_model_version.sql:8]: found a tiny bug: model version api didnt return `workspace_id`

Comment (mapmeld) [reply master/static/srv/get_model_version.sql:8]: :+1:

File: webui/react/src/pages/ModelDetails.tsx [hunk 5]
```diff
@@ -68,6 +70,10 @@ const ModelDetails: React.FC = () => {
     Loaded: (usersPagination) => usersPagination.users,
     NotLoaded: () => [],
   });
+  const ensureWorkspacesFetched = useEnsureWorkspacesFetched(canceler.current);
+  const workspaces = Loadable.getOrElse([], useWorkspaces());
+  const workspace = workspaces.find((ws) => ws.id === model?.model.workspaceId);
```
Comment (mapmeld) [inline webui/react/src/pages/ModelDetails.tsx:75]: How many workspaces are returned by `useEnsureWorkspacesFetched` - is it possible that there could be a pagination issue and it would be better to have the store retrieve the workspace by ID here?

Comment (keita-determined) [reply webui/react/src/pages/ModelDetails.tsx:75]: So far, all workspaces. 
[GetWorkspaces](http://latest-master.determined.ai:8080/docs/rest-api/#/Workspaces/GetWorkspaces) doesnt have `ids` params. 
I added `useEnsureWorkspacesFetched` here to follow the patterns somewhere else. Especially when ws name is changes, `useEnsureWorkspacesFetched` can update the data. but i agree that its not efficient. 
The solutions i can think of are 
- Remove `useEnsureWorkspacesFetched` in model and model version cuz even if ws name is updated, it wouldnt be a big issue not to reflect the change.
- Add `workspace ids` in [GetWorkspaces](http://latest-master.determined.ai:8080/docs/rest-api/#/Workspaces/GetWorkspaces)

Comment (ClaireNeveu) [reply webui/react/src/pages/ModelDetails.tsx:75]: You could add `ensureWorkspaceFetched(id: number)` and `getWorkspace(id: number): O<L<W>>` if you wanted to be defensive.

Comment (keita-determined) [reply webui/react/src/pages/ModelDetails.tsx:75]: i think it's fine to fetch all ws for now. As Alex said before, the number of ws wouldnt be a lot, and i feel like adding `ensureWorkspaceFetched` in ws store is unnecessary cuz backend can handle it. we have [a filter](https://github.com/determined-ai/determined/blob/master/proto/src/determined/common/v1/common.proto#L34) for this purpose. Currently, the filter is used in [GetExperiments](http://latest-master.determined.ai:8080/docs/rest-api/#/Experiments/GetExperiments), but we can extend it in [GetWorkspaces](http://latest-master.determined.ai:8080/docs/rest-api/#/Workspaces/GetWorkspaces).
@mapmeld what do you think?

**Comment by mapmeld** [comment]
Screen Shot 2023-02-06 at 2 39 49 PM

It seems counter-intuitive to me that "Uncategorized" links to the workspace's model registry, and "Model Registry" links to the global model registry page

**Comment by ClaireNeveu** [comment]
"Model registry" should link to the "tab" of the relevant workspace. I don't think uncategorized can really show a workspace page so we can special-case that one.

**Review by mapmeld — APPROVED** [review]
✅  LGTM
